### PR TITLE
Roe 206 filings endpoint

### PR DESCRIPTION
**Pull request by mattch1**
https://companieshouse.atlassian.net/browse/ROE-206

Added the controller and stubbed the service to map the submission.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -66,6 +66,11 @@
             <artifactId>api-sdk-manager-java-library</artifactId>
             <version>1.0.4</version>
         </dependency>
+        <dependency>
+            <groupId>uk.gov.companieshouse</groupId>
+            <artifactId>private-api-sdk-java</artifactId>
+            <version>2.0.89</version>
```
Comment (mwestacott) [inline pom.xml:72]: Latest version is 2.0.147 - https://github.com/companieshouse/private-api-sdk-java/releases/tag/2.0.147. Any reason not to go for the latest?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import uk.gov.companieshouse.overseasentitiesapi.utils.ApiLogger;
+
+import java.util.HashMap;
+
+import static uk.gov.companieshouse.overseasentitiesapi.utils.Constants.ERIC_REQUEST_ID_KEY;
+import static uk.gov.companieshouse.overseasentitiesapi.utils.Constants.OVERSEAS_ENTITY_ID_KEY;
+import static uk.gov.companieshouse.overseasentitiesapi.utils.Constants.TRANSACTION_ID_KEY;
+
+@RestController
+@RequestMapping("/transactions/{transaction_id}/overseas-entity/{overseas_entity_id}/filings")
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java:25]: Why aren't we prefixing this with 'private' like for CS? If it remains like this we'll need to update the interceptor config.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+public class FilingsController {
+
+    @Autowired
+    private FilingsService filingService;
+
+    @GetMapping
+    public ResponseEntity<FilingApi[]> getFiling(
+            @RequestAttribute("transaction") Transaction transaction,
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java:33]: There's a new request constant for "transaction" now in the Constants class. Please reference and use that here.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            @PathVariable(OVERSEAS_ENTITY_ID_KEY) String overseasEntityId,
+            @PathVariable(TRANSACTION_ID_KEY) String transactionId,
+            @RequestHeader(value = ERIC_REQUEST_ID_KEY) String requestId) {
+
+        var logMap = new HashMap<String, Object>();
+        logMap.put(TRANSACTION_ID_KEY, transactionId);
+        ApiLogger.infoContext(requestId, "Calling service to retrieve filing", logMap);
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java:40]: I don't think the log message is quite right - the service method is not retrieving a filing, it's submitting one after retrieving data from the OE submission record in Mongo.

File: spec/schema.json [hunk 5]
```diff
@@ -54,6 +54,48 @@
...
+        },
+        {
+          "name": "overseas_entity_id",
+          "in": "path",
+          "required": true,
+          "type": "string"
+        }
+      ],
+      "get": {
+        "summary": "Gets filing data ready to send send to CHIPS",
```
Comment (markpit) [inline spec/schema.json:74]: 2 'send's in summary

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java [hunk 6]
```diff
@@ -20,6 +20,7 @@ private Constants() {
     // Request attribute names
     public static final String TRANSACTION_KEY = "transaction";
 
-    // Filings
+    public static final String OVERSEAS_ENTITY_ID_KEY = "overseas_entity_id";
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:23]: This is used to reference a path variable like the TRANSACTION_ID_KEY so please move up, just under that constant on line 12.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java [hunk 7]
```diff
@@ -20,6 +20,7 @@ private Constants() {
...
+        // Filings
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:24]: Comment is too far indented.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java (hunk 2 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsController.java:25]: we'll need the /private prefix as that is what the transaction-api will be looking to call

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    @Mock
+    private ApiClientService apiClientService;
+
+    @Mock
+    private ApiClient apiClient;
+
+    private Transaction transaction;
+
+    @Test
+    void test() throws SubmissionNotFoundException, ServiceException {
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingServiceTest.java:45]: Suggest changing this method name to explain what the method is testing. So something like 'testSuccessfulFilingGeneration'.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        Optional<OverseasEntitySubmissionDto> submissionOpt = Optional.of(overseasEntitySubmissionDto);
+        when(overseasEntitiesService.getOverseasEntitySubmission(OVERSEAS_ENTITY_ID)).thenReturn(submissionOpt);
+        FilingApi filing = filingsService.generateOverseasEntityFiling(OVERSEAS_ENTITY_ID, transaction);
+        assertEquals(FILING_KIND_OVERSEAS_ENTITY, filing.getKind());
+        assertEquals(FILING_DESCRIPTION, filing.getDescriptionIdentifier());
+        assertEquals("Joe Bloggs Ltd", filing.getData().get("name"));
+    }
+
+    @Test
+    void testWhenSubmissionThrowsException()  {
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingServiceTest.java:57]: And this method 'testExceptionThrownWhenFilingGenerationFailsDueToNoOverseasEntityFound'.

Quite verbose, but reads nicely when viewing all the tests in an IDE.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    private Transaction transaction;
+
+    @InjectMocks
+    private FilingsController filingsController;
+
+    @Mock
+    private FilingsService filingsService;
+
+    @Test
+    void getFiling() throws ServiceException, SubmissionNotFoundException {
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/controller/FilingsControllerTest.java:37]: Suggest renaming these test methods - see comments in test class below. Up to you.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/OverseasEntitiesService.java [hunk 11]
```diff
@@ -89,4 +91,16 @@ private void addOverseasEntityResourceToTransaction(Transaction transaction,
         transaction.setResources(Collections.singletonMap(submissionUri, overseasEntityResource));
         transactionService.updateTransaction(transaction, passthroughTokenHeader);
     }
+
+    public Optional<OverseasEntitySubmissionDto> getOverseasEntitySubmission(String submissionId) {
+        var submission = overseasEntitySubmissionsRepository.findById(submissionId);
+        if (submission.isPresent()) {
+            ApiLogger.info(String.format("%s: Confirmation Statement Submission found. About to return", submission.get().getId()));
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/OverseasEntitiesService.java:98]: "Confirmation Statement Submission found" needs to be changed

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/OverseasEntitiesService.java [hunk 12]
```diff
@@ -89,4 +91,16 @@ private void addOverseasEntityResourceToTransaction(Transaction transaction,
...
+
+            var json = overseasEntityDtoDaoMapper.daoToDto(submission.get());
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/OverseasEntitiesService.java:100]: i'd rename json var to dto as it's not strictly json (string) but the dto object

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java (hunk 7 above)
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:24]: "// Filings" indentation needs fixing

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java (hunk 6 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:22]: @mattch1 I think you missed this? No response to my comment so not sure. Could you please action.

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:22]: Oh I see, you moved the wrong lines. TRANSACTION_KEY constant and comment should remain where it is but the new OVERSEAS_ENTITY_ID_KEY constant needs to move up to just below line 12, as it refers to a URI path attribute.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+    public FilingApi generateOverseasEntityFiling(String overseasEntityId, Transaction transaction)
+            throws SubmissionNotFoundException, ServiceException {
+        var filing = new FilingApi();
+        filing.setKind(FILING_KIND_OVERSEAS_ENTITY);
+        setFilingApiData(filing, overseasEntityId, transaction);
+        return filing;
+    }
+
+    private void setFilingApiData(FilingApi filing, String overseasEntityId, Transaction transaction) throws SubmissionNotFoundException, ServiceException {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java:39]: Yes, please address Sonar violations reported on this method and another.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java (hunk 6 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/Constants.java:22]: @mattch1 this still needs attention.

**Review by mwestacott — APPROVED** [review]

**Comment by mwestacott** [comment]
Please check with @markpit before merging.

**Review by markpit — APPROVED** [review]

**Pull request merged**